dyno-recipes/src/main/java/com/netflix/dyno/recipes/lock/DynoLockClient.java

```diff
@@ -0,0 +1,369 @@
+        this.pool = pool;
+        this.votingHostsSelector = votingHostsSelector;
+        // Threads for locking and unlocking
+        this.service = Executors.newCachedThreadPool();
+        this.quorum = votingHostsSelector.getVotingSize() / 2 + 1;
+        this.timeout = timeout;
+        this.timeoutUnit = unit;
+        // We want to release all locks in case of a graceful shutdown
+        Runtime.getRuntime().addShutdownHook(new Thread(() -> cleanup()));
+    }
+
+    public void setTimeoutUnit(TimeUnit timeoutUnit) {
+        this.timeoutUnit = timeoutUnit;
+    }
+
+    public void setTimeout(long timeout) {
+        this.timeout = timeout;
+    }
+
+    private static String getRandomString() {
+        return UUID.randomUUID().toString();
+    }
+
+    /**
+     * Gets list of resources which are locked by the client
+     * @return
+     */
+    public List<String> getLockedResources() {
+        return new ArrayList<>(resourceKeyMap.keySet());
+    }
+
+    /**
+     * Release the lock (if held) on the resource
+     * @param resource
+     */
+    public void releaseLock(String resource) {
+        if (!checkResourceExists(resource)) {
+            logger.info("No lock held on {}", resource);
+            return;
+        }
+        CountDownLatch latch = new CountDownLatch(votingHostsSelector.getVotingSize());
+        votingHostsSelector.getVotingHosts().getEntireList().stream()
+                .map(host -> new CheckAndRunHost(host, pool, "del", resource, resourceKeyMap.get(resource)))
+                .forEach(ulH -> CompletableFuture.supplyAsync(ulH, service)
+                        .thenAccept(result -> latch.countDown())
+                );
+        boolean latchValue = false;
+        try {
+            latchValue = latch.await(timeout, timeoutUnit);
+        } catch (InterruptedException e) {
+            logger.info("Interrupted while releasing the lock for resource {}", resource);
+        }
+        if (latchValue) {
+            logger.info("Released lock on {}", resource);
+        } else {
+            logger.info("Timed out before we could release the lock");
+        }
+        resourceKeyMap.remove(resource);
+    }
+
+    /**
+     * Return a timer task which will recursively schedule itself when extension is successful.
+     * @param runJob
+     * @param resource
+     * @param ttlMS
+     * @param extensionFailed
+     * @return
+     */
+    private TimerTask getExtensionTask(Timer runJob, String resource, long ttlMS, Consumer<String> extensionFailed) {
+        return new TimerTask() {
+            @Override
+            public void run() {
+                long extendedValue = extendLock(resource, ttlMS);
+                if (extendedValue > 0) {
+                    logger.info("Extended lock on {} for {} MS", resource, ttlMS);
+                    TimerTask task = getExtensionTask(runJob, resource, ttlMS, extensionFailed);
+                    runJob.schedule(task, extendedValue/2);
```

Comment:
A quick comment about how you decide when to reschedule here or in the function header comment would be useful.